src/java.base/share/classes/java/lang/VirtualThread.java

```diff
@@ -844,18 +844,32 @@ boolean joinNanos(long nanos) throws InterruptedException {
         return true;
     }
 
+    @Override
+    void blockedOn(Interruptible b) {
+        notifyJvmtiDisableSuspend(true);
+        try {
+            super.blockedOn(b);
+        } finally {
+            notifyJvmtiDisableSuspend(false);
+        }
+    }
+
     @Override
     @SuppressWarnings("removal")
     public void interrupt() {
         if (Thread.currentThread() != this) {
             checkAccess();
+
+            // if current thread is a virtual thread then prevent it from being
```

Comment:
Is the use of "current thread" here meant to imply "Thread.currentThread()"? If so, is there a check missing for `Thread.currentThread().isVirtual()` here?

Reply:
It's correct. There are two threads, the current thread that is executing the method, plus "this". Serguei is working to change notifyJvmtiDisableSuspend to be a static method (JDK-8322744) to make it clearer that it disable suspend of the current thread rather than the receiver.

Reply:
> Serguei is working to change notifyJvmtiDisableSuspend to be a static method (JDK-8322744) to make it clearer that it disable suspend of the current virtual thread rather than the receiver.

That explains it then. Before commenting, I did check if `notifyJvmtiDisableSuspend` operates on the `Thread.currentThread()` but it was an instance method which ended up being a `native` method and I didn't dig deeper but assumed it works on the thread instance on which it is called.